components/motor/gpio/loop.go

```diff
@@ -32,57 +29,73 @@ func (m *EncodedMotor) State(ctx context.Context) ([]float64, error) {
 func (m *EncodedMotor) updateControlBlock(ctx context.Context, setPoint, maxVel float64) error {
 	// Update the Trapezoidal Velocity Profile block with the given maxVel for velocity control
 	velConf := control.BlockConfig{
-		Name: "trapz",
+		Name: m.blockNames["trapz"],
 		Type: "trapezoidalVelocityProfile",
 		Attribute: rdkutils.AttributeMap{
 			"max_vel":    maxVel,
 			"max_acc":    30000.0,
 			"pos_window": 0.0,
 			"kpp_gain":   0.45,
 		},
-		DependsOn: []string{"set_point", "endpoint"},
+		DependsOn: []string{m.blockNames["constant"], m.blockNames["endpoint"]},
 	}
-	if err := m.loop.SetConfigAt(ctx, "trapz", velConf); err != nil {
+	if err := m.loop.SetConfigAt(ctx, m.blockNames["trapz"], velConf); err != nil {
 		return err
 	}
 
 	// Update the Constant block with the given setPoint for position control
 	posConf := control.BlockConfig{
-		Name: "set_point",
+		Name: m.blockNames["constant"],
 		Type: "constant",
 		Attribute: rdkutils.AttributeMap{
 			"constant_val": setPoint,
 		},
 		DependsOn: []string{},
 	}
-	if err := m.loop.SetConfigAt(ctx, "set_point", posConf); err != nil {
+	if err := m.loop.SetConfigAt(ctx, m.blockNames["constant"], posConf); err != nil {
 		return err
 	}
 	return nil
 }
 
 // validateControlConfig ensures the programmatically edited blocks are named correctly.
 func (m *EncodedMotor) validateControlConfig(ctx context.Context) error {
-	constBlock, err := m.loop.ConfigAt(ctx, "set_point")
+	m.blockNames = make(map[string]string)
+
+	// These three blocks are the only block names that are used by EncodedMotor
+	// verify constant block exits and store its name
+	constBlocks, err := m.loop.ConfigAtType(ctx, "constant")
+	m.logger.CDebugf(ctx, "const blocks = %v", constBlocks)
 	if err != nil {
-		return errConstantBlock
+		return err
 	}
-	m.logger.CDebugf(ctx, "constant block: %v", constBlock)
-	endBlock, err := m.loop.ConfigAt(ctx, "endpoint")
-	if err != nil {
-		return errEndpointBlock
+	if len(constBlocks) != 1 {
+		return errMissingBlock("constant")
 	}
-	m.logger.CDebugf(ctx, "endpoint block: %v", endBlock)
-	trapzBlock, err := m.loop.ConfigAt(ctx, "trapz")
+	m.blockNames["constant"] = constBlocks[0].Name
+
+	// verify trapz block exists and store its name
+	trapzBlocks, err := m.loop.ConfigAtType(ctx, "trapezoidalVelocityProfile")
+	m.logger.CDebugf(ctx, "trapz blocks = %v", trapzBlocks)
 	if err != nil {
-		return errTrapzBlock
+		return err
+	}
+	if len(trapzBlocks) != 1 {
+		return errMissingBlock("trapezoidalVelocityProfile")
 	}
-	m.logger.CDebugf(ctx, "trapz block: %v", trapzBlock)
-	pidBlock, err := m.loop.ConfigAt(ctx, "PID")
+	m.blockNames["trapz"] = trapzBlocks[0].Name
```

Comment:
There is so much duplicated code in here. Just make a helper function.

```go
func (m *EncodedMotor) validateControlConfig(ctx context.Context) error {
    ...

    func storeBlockOfType(type, name string) error {
        blocks, err := m.loop.ConfigAtType(ctx, type)
        if err != nil { return err }
        if len(blocks) != 1 { return errMissingBlock(type) }
        m.blockNames[name] = blocks[0].Name
        return nil
    }

    return multierr.Combine(
        storeBlockOfType("constant", "constant"),
        storeBlockOfType("trapezoidalVelocityProfile", "trapz"),
        storeBlockOfType("endpoint", "endpoint")
    )
}
```

That way, if something needs changing (e.g., the `CDebugf` lines), you only need to change it once. I wish it was impossible to copy and paste code; it copies bugs. :confounded:

Reply:
will do